### PR TITLE
Backports 0.17 pr19

### PR DESCRIPTION
**Pull request by PastaPastaPasta**

### CODE REVIEW

File: test/functional/feature_blocksdir.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env python3
+# Copyright (c) 2018 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the blocksdir option.
+"""
+
+import os
+import re
```
Comment (UdjinM6) [inline test/functional/feature_blocksdir.py:9]: Linter complains `test/functional/feature_blocksdir.py:9:1: F401 're' imported but unused`. Should either do
```suggestion
```
or (partially?) backport https://github.com/bitcoin/bitcoin/pull/12853

**Review by UdjinM6 — COMMENTED** [review]
One issue, otherwise LGTM

**Review by xdustinface — APPROVED** [review]
utACK

**Comment by UdjinM6** [comment]
LGTM but I think we should merge #3682 first in case there are some merge conflicts.

**Comment by UdjinM6** [comment]
Needs rebase

**Comment by PastaPastaPasta** [comment]
Rebased @UdjinM6

File: doc/developer-notes.md [hunk 2]
```diff
@@ -208,22 +246,73 @@ make cov
 # A coverage report will now be accessible at `./test_dash.coverage/index.html`.
 ```
 
+**Sanitizers**
+
+Bitcoin can be compiled with various "sanitizers" enabled, which add
```
Comment (UdjinM6) [inline doc/developer-notes.md:251]:
```suggestion
Dash Core can be compiled with various "sanitizers" enabled, which add
```

File: doc/developer-notes.md [hunk 3]
```diff
@@ -629,6 +721,52 @@ Current subtrees include:
...
+reason, on 64-bit Unix systems we rely on an internal LevelDB optimization that
+uses `mmap()` + `close()` to open table files without actually retaining
+references to the table file descriptors. If you are upgrading LevelDB, you must
+sanity check the changes to make sure that this assumption remains valid.
+
+In addition to reviewing the upstream changes in `env_posix.cc`, you can use `lsof` to
+check this. For example, on Linux this command will show open `.ldb` file counts:
+
+```bash
+$ lsof -p $(pidof bitcoind) |\
```
Comment (UdjinM6) [inline doc/developer-notes.md:745]:
```suggestion
$ lsof -p $(pidof dashd) |\
```

File: doc/developer-notes.md [hunk 4]
```diff
@@ -1,6 +1,43 @@
...
+        - [debug.log](#debuglog)
+        - [Testnet and Regtest modes](#testnet-and-regtest-modes)
+        - [DEBUG_LOCKORDER](#debug_lockorder)
+        - [Valgrind suppressions file](#valgrind-suppressions-file)
+        - [Compiling for test coverage](#compiling-for-test-coverage)
+    - [Locking/mutex usage notes](#lockingmutex-usage-notes)
+    - [Threads](#threads)
+    - [Ignoring IDE/editor files](#ignoring-ideeditor-files)
+- [Development guidelines](#development-guidelines)
+    - [General Bitcoin Core](#general-bitcoin-core)
```
Comment (UdjinM6) [inline doc/developer-notes.md:22]:
```suggestion
    - [General Dash Core](#general-dash-core)
```

**Review by UdjinM6 — COMMENTED** [review]
Looks good 👍 but I just noticed that we missed some dashification in docs, pls see suggestions below

**Review by UdjinM6 — APPROVED** [review]
utACK

**Pull request merged**